doc/poddoc/man/xtlpc.8
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "XTLPC 8"
.TH XTLPC 8 "2009-11-16" "GNUspool Release 1" "GNUspool Print Manager"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
xtlpc \- LPD protocol server driver
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBxtlpc\fR
\&\fI\-H hostname\fR
\&\fI\-f ctrlfile\fR
[ \fI\-F fifo\fR ]
[ \fI\-S host\fR ]
[ \fI\-P ptrname\fR ]
[ \fI\-N\fR ]
[ \fIfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBxtlpc\fR is a terminal server program for use by \fBGNUspool\fR where
the destination is an \s-1LPD\s0 protocol server. Input is usually taken from
standard input, or from the supplied file name.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-H \fIhostname\fR" 8
.IX Item "-H hostname"
This is the host name or \s-1IP\s0 address to be used to access the
printer. In \fBGNUspool\fR setup files, it may be appropriate to
substitute \f(CW$SPOOLDEV\fR or \f(CW$SPOOLPTR\fR here.
.IP "\-f \fIctrlfile\fR" 8
.IX Item "-f ctrlfile"
This specifies the location of the control file, typically
\&\fIxtlpc-ctrl\fR in the internal programs directory.
.IP "\-S \fIhostname\fR" 8
.IX Item "-S hostname"
This is the host name or \s-1IP\s0 address of the sending host, in the cases
where \fIgethostname\fR\|(3) does not yield an appropriate answer.
.IP "\-P \fIptrname\fR" 8
.IX Item "-P ptrname"
The \s-1LPD\s0 protocol supplies a name to the printer. By default, this is
taken from \fB\s-1SPOOLPTR\s0\fR, i.e. the name given as the name of the printer
under \fBGNUspool\fR. Some terminal servers require a specific name to be
given to the printer, perhaps to select which printer, or printer
option of several is to be specified. This option overrides this
printer name in the protocol.
.IP "\-F \fIfifo\fR" 8
.IX Item "-F fifo"
Take input from the named \fIfifo\fR rather than standard input or the
supplied file.
.Sp
This is useful in situations where the \s-1LPD\s0 protocol has to be run from
a command line, i.e. a console session, and files sent to it by having
\&\fBGNUspool\fR write to the fifo.
.IP "\-N" 8
.IX Item "-N"
Avoid sending zero length files, which confuse some servers.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
\&\fBxtlpc\fR generates any appropriate diagnostics on standard error.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
Normal termination is denoted by an exit code of zero.
.IP "1" 4
.IX Item "1"
Usage error, invalid option etc.
.IP "2" 4
.IX Item "2"
No control file given.
.IP "3" 4
.IX Item "3"
No destination host given.
.IP "4" 4
.IX Item "4"
Cannot open input file.
.IP "5" 4
.IX Item "5"
Unknown host name.
.IP "6" 4
.IX Item "6"
Cannot open control file.
.IP "7" 4
.IX Item "7"
Syntax errors in control file.
.IP "8" 4
.IX Item "8"
No spool directory defined.
.IP "9" 4
.IX Item "9"
Cannot select spool directory.
.IP "10" 4
.IX Item "10"
Control card format not given in control file
.IP "11" 4
.IX Item "11"
Invalid format control card in control file
.IP "12" 4
.IX Item "12"
Cannot write to spool directory
.IP "13" 4
.IX Item "13"
Cannot create spool file
.IP "14" 4
.IX Item "14"
Cannot allocate file name.
.IP "15" 4
.IX Item "15"
Cannot open \s-1FIFO\s0.
.IP "16" 4
.IX Item "16"
Name given is not a \s-1FIFO\s0.
.IP "17" 4
.IX Item "17"
Cannot specify input and \s-1FIFO\s0.
.IP "20" 4
.IX Item "20"
Cannot create sequence file
.IP "30" 4
.IX Item "30"
No port name in control file
.IP "31" 4
.IX Item "31"
\&\s-1TCP\s0 protocol missing
.IP "32" 4
.IX Item "32"
Cannot find named port in service file
.IP "33" 4
.IX Item "33"
Cannot open socket.
.IP "34" 4
.IX Item "34"
Cannot bind port
.IP "35" 4
.IX Item "35"
Cannot allocate \s-1TCP\s0 device
.IP "36" 4
.IX Item "36"
Cannot make connection
.IP "37" 4
.IX Item "37"
Cannot allocate \s-1TCP\s0 space or change stream
.IP "40" 4
.IX Item "40"
Write failure writing to socket
.IP "41" 4
.IX Item "41"
Lost spool file
.IP "42" 4
.IX Item "42"
Read failure ack packet.
.IP "43" 4
.IX Item "43"
Error return from remote.
.IP "50" 4
.IX Item "50"
Cannot determine own host name
.IP "51" 4
.IX Item "51"
Invalid host name
.IP "200" 4
.IX Item "200"
Exit on signal.
.SH "FILES"
.IX Header "FILES"
\&\fIxtlpc-ctrl\fR
control file
.PP
\&\fI/var/spool/gnuspool/xtlpc\fR
spool directory (may be changed in \fIxtlpc-ctrl\fR)
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIxilp\fR\|(8),
\&\fIxtelnet\fR\|(8),
\&\fIxtftp\fR\|(8),
\&\fIxtlhp\fR\|(8).
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2009 Free Software Foundation, Inc.
This is free software. You may redistribute copies of it under the
terms of the \s-1GNU\s0 General Public License
<http://www.gnu.org/licenses/gpl.html>.
There is \s-1NO\s0 \s-1WARRANTY\s0, to the extent permitted by law.
.SH "AUTHOR"
.IX Header "AUTHOR"
John M Collins, Xi Software Ltd.


debian/changelog
gnuspool (1.7) unstable; urgency=low

  * strip out unneeded sections of postinst etc. and do some related tidying
  * remove MSWIN subdirectory and bump version in configure

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Mon, 26 Apr 2010 12:26:21 +0000

gnuspool (1.6) unstable; urgency=low

  [ John M Collins (Xi Software Ltd) ]
  * Initial release (Closes: #556996)

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Tue, 17 Nov 2009 10:16:21 +0000


src/lpdint/xtlpc.c
/* xtlpc.c -- driver to output using LPD protocol

   Copyright 2008 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"

#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#ifdef	HAVE_FCNTL_H
#include <fcntl.h>
#endif
#include <errno.h>
#ifdef	HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#include "defaults.h"
#include "files.h"
#include "incl_unix.h"
#include "incl_net.h"
#include "incl_sig.h"
#include "lpctypes.h"
#include "network.h"
#include "spq.h"
#include "xfershm.h"
#include "q_shm.h"

#define	PIDFILE	".xtlpcpid"

static	char	*myhostname;
netid_t	myhostid;
static	int	nonull;

int		loops = 3, unresport = 0;
int		had_alarm = 0;
unsigned	loopwait = 1;
float		lingertime = 0;
unsigned	input_timeout = 5, output_timeout = 5, send_retries = 0;

extern	uid_t	Realuid, Effuid, Daemuid;

extern int	parsecf(FILE *);
extern int	evalcond(struct condition *);

RETSIGTYPE	cleanupfiles(int signum)
{
	struct	varname	*vp;

	vp = lookuphash("CFILE");
	if  (vp->vn_value  && vp->vn_value[0])
		unlink(vp->vn_value);
	vp = lookuphash("DFILE");
	if  (vp->vn_value  && vp->vn_value[0])
		unlink(vp->vn_value);
	unlink(PIDFILE);
	if  (signum != 0)
		exit(200);
}

RETSIGTYPE	alarmhandler(int signum)
{
#ifdef	UNSAFE_SIGNALS
	signal(signum, alarmhandler);
#endif
	had_alarm++;
}

static netid_t	host_by_nameoraddr(char *str)
{
	if  (isdigit(str[0]))  {
		netid_t  ina;
#ifdef	DGAVIION
		struct	in_addr	ina_str;
		ina_str = inet_addr(str);
		ina = ina_str.s_addr;
#else
		ina = inet_addr(str);
#endif
		if  (ina == -1L)
			return  (netid_t) 0;
		return  ina;
	}
	else  {
		struct  hostent  *hp = gethostbyname(str);
		return  hp? *(netid_t *) hp->h_addr: (netid_t) 0;
	}
}

static void	envassign(char *name, char *defvalue)
{
	struct	varname	*vp;

	vp = lookuphash(name);
	if  (vp->vn_value)  {
		if  (vp->vn_value[0])
			return;
		free(vp->vn_value);
	}
	vp->vn_value = stracpy(defvalue); /* Must be free-able */
}

static void	assignenvs(char *sendhost)
{
	extern	char	**environ;
	char	**ep;
	struct	hostent	*hp;
	char	myhost[256];

	for  (ep = environ;  *ep;  ep++)  {
		char	*ec, *eqp;
		struct	varname	*vp;
		ec = stracpy(*ep);
		if  ((eqp = strchr(ec, '=')))  {
			*eqp = '\0';
			vp = lookuphash(ec);
			if  (vp->vn_value)
				free(vp->vn_value);
			vp->vn_value = stracpy(eqp+1);
		}
		free(ec);
	}

	if  (sendhost)  {
		if  (!(myhostid = host_by_nameoraddr(sendhost)))  {
			fprintf(stderr, "Invalid host name %s\n", sendhost);
			exit(51);
		}
		if  (isdigit(sendhost[0]) &&  (hp = gethostbyaddr((char *) &myhostid, sizeof(myhostid), AF_INET)))
			strncpy(myhost, hp->h_name, sizeof(myhost)-1);
		else
			strncpy(myhost, sendhost, sizeof(myhost)-1);
		myhost[sizeof(myhost) - 1] = '\0';
	}
	else  {
		gethostname(myhost, sizeof(myhost));
		if  (!(hp = gethostbyname(myhost)))  {
			fprintf(stderr, "Who am I?\n");
			exit(50);
		}
		myhostid = *(netid_t *) hp->h_addr;
	}
	myhostname = stracpy(myhost);
}

static void	reset_envirvar(char *name, char *value)
{
	struct	varname	*vp = lookuphash(name);
	if  (vp->vn_value)
		free(vp->vn_value);
	vp->vn_value = stracpy(value);
}

FILE  *generate_cdfname(int startseq, char *vname, char **oldval)
{
	struct	varname	*vp = lookuphash(vname);
	int	ch, cnt;
	FILE	*result;

	if  (!vp->vn_value || vp->vn_value[0] == '\0')  {
		fprintf(stderr, "%s not defined in control file\n", vname);
		exit(10);
	}
	if  ((int) strlen(vp->vn_value) > 70 || strchr(vp->vn_value, '/'))  {
		fprintf(stderr, "Invalid value \'%s\' for %s\n", vp->vn_value, vname);
		exit(11);
	}
	for  (ch = 'A';  ch <= 'Z';  ch++)  {
		for  (cnt = startseq;  cnt < 1000;  cnt++)  {
			int	fd;
			char	outname[160];
			sprintf(outname, vp->vn_value, ch, cnt);
			if  ((fd = open(outname, O_CREAT|O_EXCL|O_WRONLY, 0666)) < 0)  {
				if  (errno == EACCES)  {
					fprintf(stderr, "No write access to spool directory\n");
					exit(12);
				}
				continue;
			}
			if  (!(result = fdopen(fd, "w")))  {
				fprintf(stderr, "Trouble opening %s\n", outname);
				exit(13);
			}
			*oldval = vp->vn_value;
			vp->vn_value = stracpy(outname);
			return  result;
		}
		startseq = 1;
	}
	fprintf(stderr, "Unable to allocate a file name for %s\n", vname);
	exit(14);
}

void	restore_cdfname(char *vname, char *oldval)
{
	struct  varname  *vp = lookuphash(vname);
	if  (vp->vn_value)  {
		unlink(vp->vn_value);
		free(vp->vn_value);
	}
	vp->vn_value = oldval;
}

static char *	conv_line(char *str, int *length)
{
	char	*result = malloc((unsigned) (strlen(str) + 10));
	char	*cp, *rp;

	if  (!result)
		nomem();
	rp = result;
	cp = str;

	while  (*cp)  {
		if  (*cp == '^')  {
			if  (*++cp == '^')
				*rp++ = '^';
			else
				*rp++ = *cp++ & 0x1f;
		}
		else  if  (*cp == '%')  {
			char	*np;
			struct	stat	sbuf;
			char	nbuf[20];
			do  cp++;
			while  (*cp == ' ');
			if  (!(np = strchr(cp, ' ')))
				continue;
			*np = '\0';
			if  (stat(cp, &sbuf) < 0)  {
				*np = ' ';
				continue;
			}
			*np = ' ';
			cp = np;
			sprintf(nbuf, "%ld", nonull && sbuf.st_size == 0? 1: sbuf.st_size);
			np = nbuf;
			do  *rp++ = *np++;
			while  (*np);
		}
		else
			*rp++ = *cp++;
	}

	*rp = '\0';
	*length = rp - result;
	return  result;
}

static int	getseq(void)
{
	int	fd, result;
	char	buf[100];

	if  ((fd = open(".seq", O_RDWR)) < 0)  {
		if  ((fd = open(".seq", O_RDWR|O_CREAT|O_TRUNC, 0666)) < 0)  {
			fprintf(stderr, "Cannot create sequence file\n");
			exit(20);
		}
		result = 1;
	}
	else  {
		int	inb = read(fd, buf, sizeof(buf));
		if  (inb <= 1)
			result = 1;
		result = atoi(buf) + 1;
		lseek(fd, 0L, 0);
	}
	sprintf(buf, "%d\n", result);
	write(fd, buf, strlen(buf));
	close(fd);
	return  result;
}

static void	savepid(void)
{
	int	fd;
	char	buf[40];

	if  ((fd = open(PIDFILE, O_RDWR|O_CREAT|O_TRUNC, 0666)) < 0)
		return;
	sprintf(buf, "%d\n", getpid());
	write(fd, buf, strlen(buf));
	close(fd);
}

/* Set up network stuff */

static int	init_network(netid_t hostid)
{
	int	qsock;
	struct	protoent  *pp;
	struct	varname	*vp;
	char	*tcp_protoname;
	short	qportnum, resport;
	struct	sockaddr_in	server, client;
	short	tcpproto;

	/* First get port name/number */

	vp = lookuphash("PORTNAME");
	if  (!vp->vn_value  ||  !vp->vn_value[0])  {
		fprintf(stderr, "No port name/number defined in control file - aborting\n");
		exit(30);
	}

	/* Get TCP protocol name */

	if  (!((pp = getprotobyname("tcp"))  || (pp = getprotobyname("TCP"))))  {
		fprintf(stderr, "Cannot find tcp proto - aborting\n");
		exit(31);
	}
	tcp_protoname = pp->p_name;
	tcpproto = pp->p_proto;
	if  (isdigit(vp->vn_value[0]))
		qportnum = htons(atoi(vp->vn_value));
	else  {
		struct	servent	*sp;
		if  (!(sp = getservbyname(vp->vn_value, tcp_protoname)))  {
			fprintf(stderr, "Cannot find port %s - aborting\n", vp->vn_value);
			exit(32);
		}
		qportnum = sp->s_port;
		endservent();
	}

	endprotoent();

	resport = IPPORT_RESERVED - 4;
	BLOCK_ZERO(&server, sizeof(server));
	BLOCK_ZERO(&client, sizeof(client));
	server.sin_family = client.sin_family = AF_INET;
	server.sin_port = qportnum;
	client.sin_port = htons(resport);
	server.sin_addr.s_addr = hostid;
	client.sin_addr.s_addr = myhostid;

	if  ((qsock = socket(PF_INET, SOCK_STREAM, tcpproto)) < 0)  {
		fprintf(stderr, "Cannot open socket\n");
		exit(33);
	}
	if  (!unresport)  {
		while   (bind(qsock, (struct sockaddr *) &client, sizeof(client)) < 0)  {
			if  (errno == EACCES)  {
				fprintf(stderr, "No privilege to bind port\n");
				exit(34);
			}
			if  (--resport <= 0)  {
				fprintf(stderr, "Run out of space to bind ports\n");
				exit(37);
			}
			client.sin_port = htons(resport);
		}
	}

	do  {
		if  (connect(qsock, (struct sockaddr *) &server, sizeof(server)) >= 0)
			goto  doneconn;
		sleep(loopwait);
	}  while  (--loops > 0);

	fprintf(stderr, "Cannot connect socket\n");
	close(qsock);
	exit(36);

 doneconn:

	if  (lingertime != 0.0)  {
		struct	linger	soko;
		soko.l_linger = (int) (lingertime * 100.0 + .5);
		soko.l_onoff = 1;
		setsockopt(qsock, SOL_SOCKET, SO_LINGER, &soko, sizeof(soko));
	}

	return  qsock;
}

int	pushout(int sockfd, char * buf, int buflen)
{
	int	outb;

	while  (buflen > 0)  {
		alarm(output_timeout);
		if  ((outb = write(sockfd, buf, buflen)) < 0)  {
			if  (errno == EINTR)  {
				fprintf(stderr, "Output timeout\n");
				return  1;
			}
			fprintf(stderr, "Lost connection to remote\n");
			cleanupfiles(0);
			exit(40);
		}
		alarm(0);
		buflen -= outb;
		buf += outb;
	}
	return  0;
}

int	pushfile(int sockfd, char *filename)
{
	int	infd, inb;
	long	had = 0;
	char	outbuf[1024];

	if  ((infd = open(filename, O_RDONLY)) < 0)  {
		fprintf(stderr, "Lost file %s??\n", filename);
		cleanupfiles(0);
		exit(41);
	}
	while  ((inb = read(infd, outbuf, sizeof(outbuf))) > 0)  {
		had += inb;
		if  (pushout(sockfd, outbuf, inb))
			return  1;
	}
	if  (nonull  &&  had <= 0)  {
		outbuf[0] = '\0';
		if  (pushout(sockfd, outbuf, 1))
			return  1;
	}
	close(infd);
	return  0;
}

int	pullin(int sockfd, char * cbuf, int cbuflen)
{
	char	inbuf;

	while  (cbuflen > 0)  {
		int	nb;
		alarm(input_timeout);
		if  ((nb = read(sockfd, &inbuf, sizeof(inbuf))) <= 0)  {
			if  (nb == 0)
				fprintf(stderr, "EOF\n");
			else  {
				if  (errno == EINTR)  {
					fprintf(stderr, "Input timeout\n");
					return  1;
				}
				perror("Ack read");
			}
			fprintf(stderr, "Trouble reading ack packet\n");
			exit(42);
		}
		alarm(0);
		if  (inbuf != *cbuf)  {
			fprintf(stderr, "Error returned from remote - %d\n", inbuf);
			exit(43);
		}
		cbuflen--;
		cbuf++;
	}
	return  0;
}

MAINFN_TYPE	main(int argc, char **argv)
{
	struct	ctrltype	*cp;
	int	ch, sequence, sock, pass1 = 0;
	char	*ctrlfile = (char *) 0;
	char	*outhost = (char *) 0;
	char	*sendhost = (char *) 0;
	char	*fifoname = (char *) 0;
	char	*printername = (char *) 0;
	char	*save_cfile = (char *) 0,
		*save_dfile = (char *) 0;
	netid_t	outhostid;
	FILE	*wotfile, *cfile, *dfile, *ff = (FILE *) 0;
	struct	varname	*vp;
	extern	int	optind;
	extern	char	*optarg;
#ifdef	STRUCT_SIG
	struct	sigstruct_name  zc;
#endif
	Realuid = getuid();

	versionprint(argv, "$Revision: 1.2 $", 1);

	while  ((ch = getopt(argc, argv, "f:H:S:NF:P:l:L:s:UI:O:R:")) != EOF)
		switch  (ch)  {
		case  '?':
			fprintf(stderr, "Usage: %s -f Ctrlfile -H Outhost\n", argv[0]);
			return  1;
		case  'f':
			ctrlfile = optarg;
			break;
		case  'F':
			fifoname = optarg;
			break;
		case  'H':
			outhost = optarg;
			break;
		case  'S':
			sendhost = optarg;
			break;
		case  'P':
			printername = optarg;
			break;
		case  'N':
			nonull++;
			break;
		case  'l':
			loops = atoi(optarg);
			break;
		case  'L':
			loopwait = (unsigned) atoi(optarg);
			break;
		case  's':
			lingertime = atof(optarg);
			break;
		case  'U':
			unresport = 1;
			break;
		case  'I':
			input_timeout = (unsigned) atoi(optarg);
			break;
		case  'O':
			output_timeout = (unsigned) atoi(optarg);
			break;
		case  'R':
			send_retries = (unsigned) atoi(optarg);
			break;
		}

	if  (!ctrlfile)  {
		fprintf(stderr, "%s: You didn\'t give a control file\n", argv[0]);
		return  2;
	}
	if  (!outhost)  {
		fprintf(stderr, "%s: You didn\'t give a destination host\n", argv[0]);
		return  3;
	}

	if  (argv[optind])  {
		if  (fifoname)  {
			fprintf(stderr, "%s: You cannot specify a FIFO and an input file\n", argv[0]);
			return  17;
		}
		if  (!freopen(argv[optind], "r", stdin))  {
			fprintf(stderr, "Cannot reopen stdin\n");
			return  4;
		}
	}

	if  (!(outhostid = host_by_nameoraddr(outhost)))  {
		fprintf(stderr, "%s: Unknown host name %s\n", argv[0], outhost);
		return  5;
	}

	if  (!(wotfile = fopen(ctrlfile, "r")))  {
		fprintf(stderr, "%s: cannot open %s\n", argv[0], ctrlfile);
		return  6;
	}

	assignenvs(sendhost);

	/* Assign parameters if not defined by Xi-Text in the environment */

	envassign("SPOOLPTR", "lp");
	envassign("SPOOLHDR", "stdin");
	envassign("SPOOLUSER", "0");
	envassign("SPOOLJUNAME", "root");
	envassign("SPOOLPUNAME", "root");
	envassign("SPOOLFORM", "standard");
	envassign("SPOOLJOB", "0");
	envassign("SPOOLHOST", myhostname);
	envassign("SPOOLCPS", "1");

	/* Set up my names. Ignore the environment */

	reset_envirvar("MYHOST", myhostname);
	reset_envirvar("DESTHOST", outhost);

	/* Possibly override printer name */

	if  (printername)
		reset_envirvar("SPOOLPTR", printername);

	if  (parsecf(wotfile))  {
		fprintf(stderr, "Aborting due to control file errors\n");
		return  7;
	}
	fclose(wotfile);

	vp = lookuphash("XTLPCSPOOL");
	if  (!vp->vn_value)  {
		fprintf(stderr, "No client spool directory defined\n");
		return  8;
	}

	if  (chdir(vp->vn_value) < 0)  {
		fprintf(stderr, "Cannot select spool directory\n");
		return  9;
	}

	/* Look for fifo now if applicable after we've changed dir.  */

	if  (fifoname)  {
		struct	stat	sbuf;
		if  (stat(fifoname, &sbuf) < 0)  {
			fprintf(stderr, "%s: cannot find FIFO %s\n", argv[0], fifoname);
			return  15;
		}
		if  ((sbuf.st_mode & S_IFMT) != S_IFIFO)  {
			fprintf(stderr, "%s: %s is not a FIFO\n", argv[0], fifoname);
			return  16;
		}
		/* Turn ourself into a child process */
		if  (fork() != 0)
			return  0;
	}
	savepid();

#ifdef	STRUCT_SIG
	zc.sighandler_el = cleanupfiles;
	sigmask_clear(zc);
	zc.sigflags_el = SIGVEC_INTFLAG;
	sigact_routine(SIGTERM, &zc, (struct sigstruct_name *) 0);
	sigact_routine(SIGINT, &zc, (struct sigstruct_name *) 0);
	sigact_routine(SIGQUIT, &zc, (struct sigstruct_name *) 0);
	sigact_routine(SIGHUP, &zc, (struct sigstruct_name *) 0);
	zc.sighandler_el = alarmhandler;
	sigact_routine(SIGALRM, &zc, (struct sigstruct_name *) 0);
#else
	signal(SIGTERM, cleanupfiles);
	signal(SIGINT, cleanupfiles);
	signal(SIGQUIT, cleanupfiles);
	signal(SIGHUP, cleanupfiles);
	signal(SIGALRM, cleanupfiles);
#endif

	sock = init_network(outhostid);

	for  (;;)  {

		sequence = getseq();
		cfile = generate_cdfname(sequence, "CFILE", &save_cfile);
		dfile = generate_cdfname(sequence, "DFILE", &save_dfile);

		/* Copy stdin, or from FIFO to dfile */

		if  (fifoname)  {
			if  (!(ff = fopen(fifoname, "r")))  {
				cleanupfiles(0);
				return  0;
			}
			while  ((ch = getc(ff)) != EOF)
				putc(ch, dfile);
		}
		else  while  ((ch = getchar()) != EOF)
			putc(ch, dfile);

		fclose(dfile);

		for  (cp = card_list;  cp;  cp = cp->ctrl_next)  {
			char	*str;
			int	repnum = 1, cnt;
			struct	condition  *cc = cp->ctrl_cond;

			if  (cc  &&  cc->cond_type != COND_MULT  &&  !evalcond(cc))
				continue;

			str = expandvars(cp->ctrl_string);

			if  (cc  &&  cc->cond_type == COND_MULT  &&  cc->cond_var  &&  cc->cond_var->vn_value)
				repnum = atoi(cc->cond_var->vn_value);

			for  (cnt = 0;  cnt < repnum;  cnt++)
				fprintf(cfile, "%s\n", str);
			free(str);
		}

		fclose(cfile);

	restart:

		for  (cp = proto_list;  cp;  cp = cp->ctrl_next)  {
			char	*str = expandvars(cp->ctrl_string), *nstr;
			int	linelen;

			switch  (cp->ctrl_action)  {
			default:
				fprintf(stderr, "Unknown action???\n");
				break;

			case  CT_SENDLINEONCE:
				if  (pass1)
					break;
			case  CT_SENDLINE:
				nstr = conv_line(str, &linelen);
				if  (pushout(sock, nstr, linelen))
					goto  io_error;
				free(nstr);
			case  CT_NOP:
				break;

			case  CT_RECVLINEONCE:
				if  (pass1)
					break;
			case  CT_RECVLINE:
				nstr = conv_line(str, &linelen);
				if  (pullin(sock, nstr, linelen))
					goto  io_error;
				free(nstr);
				break;

			case  CT_SENDFILE:
				if  (!pushfile(sock, str))
					break;
			io_error:
				if  (send_retries == 0)  {
					fprintf(stderr, "Aborting due to network I/O timeouts\n");
					return  201;
				}
				send_retries--;
				fprintf(stderr, "Retrying....\n");
				close(sock);
				pass1 = 0; /* For FIFO case */
				sock = init_network(outhostid);
				goto  restart;
			}
			free(str);
		}

		if  (!fifoname)
			break;

		pass1++;

		/* Restore the CFILE/DFILE variables to their original state */

		restore_cdfname("DFILE", save_dfile);
		restore_cdfname("CFILE", save_cfile);
		if  (ff)
			fclose(ff);
	}
	cleanupfiles(0);

	return  0;
}
